hail/src/main/scala/is/hail/expr/ir/IRBuilder.scala

```diff
@@ -0,0 +1,165 @@
+
+    def renameGlobals(globalMap: Map[String, String]): TableIR =
+      rename(Map.empty, globalMap)
+
+    def filter(ir: IRProxy): TableIR =
+      TableFilter(tir, ir(env))
+  }
+
+  class IRProxy(val ir: E => IR) extends AnyVal with Dynamic {
+    def <=: (s: Symbol): BindingProxy = BindingProxy(s, this)
+
+    def apply(idx: IRProxy): IRProxy = (env: E) =>
+      ArrayRef(ir(env), idx(env))
+
+    def selectDynamic(field: String): IRProxy = (env: E) =>
+      GetField(ir(env), field)
+
+    def unary_! = new IRProxy((env: E) => ApplyUnaryPrimOp(Bang(), ir(env)))
+
+    def apply(lookup: Symbol): IRProxy = (env: E) => {
+      val eval = ir(env)
+      eval.typ match {
+        case _: TStruct =>
+          GetField(eval, lookup.name)
+        case _: TArray =>
+          ArrayRef(ir(env), ref(lookup)(env))
+      }
+    }
+
+    def typecheck(t: Type): IRProxy = (env: E) => {
+      val eval = ir(env)
+      TypeCheck(eval, env, None)
+      assert(eval.typ == t, t._toPretty + " " + eval.typ._toPretty)
+      eval
+    }
+
+    def insertFields(fields: (Symbol, IRProxy)*): IRProxy = (env: E) =>
+      InsertFields(ir(env), fields.map { case (s, fir) => (s.name, fir(env)) })
+
+    def selectFields(fields: String*): IRProxy = (env: E) =>
+      SelectFields(ir(env), fields)
+
+    def dropFields(fields: Symbol*): IRProxy = (env: E) => {
+      val struct = ir(env)
+      val typ = struct.typ.asInstanceOf[TStruct]
+      SelectFields(struct, typ.fieldNames.diff(fields.map(_.name)))
+    }
+
+    def len: IRProxy = (env: E) => ArrayLen(ir(env))
+
+    def filter(pred: LambdaProxy): IRProxy = (env: E) => {
+      val array = ir(env)
+      val eltType = array.typ.asInstanceOf[TArray].elementType
+      ArrayFilter(array, pred.s.name, pred.body(env.bind(pred.s.name -> eltType)))
+    }
+
+    def map(f: LambdaProxy): IRProxy = (env: E) => {
+      val array = ir(env)
+      val eltType = array.typ.asInstanceOf[TArray].elementType
+      ArrayMap(ir(env), f.s.name, f.body(env.bind(f.s.name -> eltType)))
+    }
+
+    def groupByKey: IRProxy = (env: E) => GroupByKey(ir(env))
+
+    def toArray: IRProxy = (env: E) => ToArray(ir(env))
+
+    private[ir] def apply(env: E): IR = ir(env)
+  }
+
+  class LambdaProxy(val s: Symbol, val body: IRProxy)
+
+  class SymbolProxy(val s: Symbol) extends AnyVal {
+    def ~> (body: IRProxy): LambdaProxy = new LambdaProxy(s, body)
+  }
+
+  case class BindingProxy(s: Symbol, value: IRProxy)
+
+  object LetProxy {
+    def bind(bindings: Seq[BindingProxy], body: IRProxy, env: E): IR =
+      bindings match {
+        case BindingProxy(sym, binding) +: rest =>
+          val name = sym.name
+          val value = binding(env)
+          Let(name, value, bind(rest, body, env.bind(name -> value.typ)))
+        case Seq() =>
+          body(env)
+      }
+  }
+
+  object let extends Dynamic {
+    def applyDynamicNamed(method: String)(args: (String, IRProxy)*): LetProxy = {
+      assert(method == "apply")
```

Comment:
An apply method would have a fixed signature. I want to be able to write `let (row = ???, global = ???) in ???`, with arbitrary keyword arguments. I was accomplishing the same thing before with something like `def apply(bindings: BindingProxy*)`, which with the `<=:` operator above let me write `let ('row <=: ???, 'glo <=: ???) in ???`, but this is cleaner.

Reply:
With `def apply(args: (String, IRProxy)*)`, you would have to write `let ("row" -> ???, "globals" -> ???)`. If I changed it to take symbols, that could become `let ('row -> ???, 'globals -> ???)`, which is maybe preferable for being less magic. What `applyDynamicNamed` is doing is hijacking the normal argument passing semantics, and giving the method body the list `args: Seq[(String, IRProxy)]`, where the pairs are constructed from the named arguments passed to the method.

Since this was the only helpful use I found for `Dynamic`, and it's not that much better than the alternative, I think I'll get rid of it to avoid the dependency and make the code clearer.